Lift inductor lowerings for jagged <-> padded dense kernels

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121947
* #126552
* #125947
* #122836
* #128057
* #127007
* __->__ #125968

This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

jbschlosser committed:
Lift inductor lowerings for jagged <-> padded dense kernels

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser commented:
gentle ping @eellison - are you the right person to review this or can you suggest someone who can? These are simply lifted from the internal fbcode lowerings and applied to the lifted ATen ops added in #125946

davidberard98 commented:
I can take a look later today

davidberard98 commented on torch/_meta_registrations.py:
```diff
@@ -6034,6 +6034,43 @@ def meta_local_scalar_dense(self: Tensor):
+        assert shape_env is not None
+        total_L = shape_env.create_unbacked_symint()
+        torch.fx.experimental.symbolic_shapes._constrain_range_for_size(
+            total_L, min=1, max=None
```
I don't think this is necessarily true is it? You could technically have length of 0 for every batch I think

jbschlosser replied:
IIRC I had to specifically set this to >= 1 to circumvent some error - let me see if I can get away with >= 0 though. I do agree that's conceptually what we want.

jbschlosser replied:
Okay cool, I was wrong about the >= 1. For posterity, this is the assert that trips when the range is unconstrained:
https://github.com/pytorch/pytorch/blob/29081059b6534377698db4a1086d745c22f2f6eb/torch/fx/experimental/symbolic_shapes.py#L5328

Some testing shows that constraining to >= 0 works fine to avoid tripping this assert.

davidberard98 commented on torch/_inductor/fbgemm_lowerings.py:
```diff
@@ -0,0 +1,342 @@
+
+
+# pyre-ignore[2,3]
+def jagged_mm_options(config, layout):
```
nit: also not used right now AFAIK

davidberard98 commented on torch/_inductor/fbgemm_lowerings.py:
```diff
@@ -0,0 +1,342 @@
+    )
+
+
+def make_offset_generator(
```
also not used I think

davidberard98 reviewed (approved):
Generally LGTM! A few nits about dead code; also, should we be using `_constrain_range_for_size` with `min=1`? It seems possible to have a total sum of lengths of 0

eellison commented on torch/_inductor/fbgemm_lowerings.py:
```diff
@@ -0,0 +1,342 @@
+        ranges=[jagged_len],
+    )
+
+    if realize:
+        # "freeze" the node so that it doesn't get inlined downstream.
+        inverse_offsets.realize()
```
Why do need to set this ?

davidberard98 replied:
**TL;DR**: the inverse_offsets code does a binary search that (at least in the past) triton was pretty bad at generating. Additionally, we typically would need to recompute (or re-lookup) the same inverse offset result many times; so instead of recomputing this a lot of times (which triton is pretty bad at), we write it to global memory and then load it when we need it. Adnan Akhundov and I did some benchmarks a while back and found that this was actually a lot faster than trying to fuse it.

Details below:

**inverse_offset computation is duplicated / used multiple times**
the logic for handling dense->jagged conversions is something like this:
* spawn a thread for each element of the output in the jagged format
* **lookup or compute the inverse of the offsets tensor** so we know how to index into dense tensor
* use that index to load your data from the dense tensor and do whatever computation you want

**lookup or compute** - Note that in the typical use case where we're trying to convert to a jagged tensor of shape [J, M], then for each element [j, m] for a given j, the inverse offset computation will be equivalent. So you'll either need to load the same data element multiple times (which is cheap because we're not loading very much data) or duplicate the computation (which is expensive for this particular kernel)

**Second - why the computation is slow**
The inverse offset calculation is done with a binary search.

On one hand, this is expected to be kind of terrible because the entire function is just branching and uncoalesced reads. On the other hand, I'm not really sure why this is worse in triton than in CUDA

eellison reviewed (commented):
deferring review to @davidberard98 who knows this code better

jbschlosser commented on torch/_inductor/fbgemm_lowerings.py:
```diff
@@ -0,0 +1,342 @@
+
+
+# pyre-ignore[3]
+def jagged_filtered_configs(
```
good catch, removed

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser committed:
Update on "Lift inductor lowerings for jagged <-> padded dense kernels"


This PR lifts internal lowerings written for FBGEMM kernels that do jagged <-> padded dense conversions. In particular, this PR provides lowerings and meta registrations for the following ATen ops:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`
    * NB: if `total_L` is not provided, the output shape is data-dependent. An unbacked SymInt is used for this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
